indigo/bulk_creator.py

```diff
@@ -0,0 +1,203 @@
+                info['amends'] = row.get('amends') or None
+                info['commencement_date'] = row.get('commencement_date') or None
+
+            except Work.DoesNotExist:
+                work = Work()
+
+                work.frbr_uri = frbr_uri
+                work.country = view.country
+                work.locality = view.locality
+                work.title = row.get('title')
+                work.publication_name = row.get('publication_name')
+                work.publication_number = row.get('publication_number')
+                work.publication_date = row.get('publication_date')
+                work.commencement_date = row.get('commencement_date')
+                work.assent_date = row.get('assent_date')
+                work.stub = not row.get('principal')
+                work.created_by_user = view.request.user
+                work.updated_by_user = view.request.user
+
+                try:
+                    work.full_clean()
+                    work.save_with_revision(view.request.user)
+
+                    # signals
+                    work_changed.send(sender=work.__class__, work=work, request=view.request)
+
+                    # info for links
+                    pub_doc_params = {
+                        'date': row.get('publication_date'),
+                        'number': work.publication_number,
+                        'publication': work.publication_name,
+                        'country': view.country.place_code,
+                        'locality': view.locality.code if view.locality else None,
+                    }
+                    info['params'] = pub_doc_params
+
+                    work_links = ['commenced_by', 'amends', 'repealed_by', 'primary_work']
+                    for link in work_links:
+                        info[link] = row.get(link)
+
+                    info['status'] = 'success'
+                    info['work'] = work
+
+                except ValidationError as e:
+                    info['status'] = 'error'
+                    if hasattr(e, 'message_dict'):
+                        info['error_message'] = ' '.join(
+                            ['%s: %s' % (f, '; '.join(errs)) for f, errs in e.message_dict.items()]
+                        )
+                    else:
+                        info['error_message'] = e.message
+
+        return works
+
+    def validate_row(self, view, row):
+        row_country = row.get('country')
+        row_locality = row.get('locality')
+        row_subtype = row.get('subtype')
+        available_subtypes = [s.abbreviation for s in Subtype.objects.all()]
+
+        row_data = row
+        row_data['country'] = view.country.code
+        row_data['locality'] = view.locality.code if view.locality else None
+        form = self.get_row_validation_form(row_data)
+
+        # Extra validation
+        # - if the subtype hasn't been registered
+        if row_subtype and row_subtype not in available_subtypes:
```

Comment:
The form could do this for you if you use a `clean_subtype` methods (put your logic in that method on the form, it'll be called automatically), or use a ChoiceField and setup the choices in the form's `__init__` method.

Ditto for the other fields -- moving the validation logic entirely into the form makes it simpler to adjust/override.

But none of this is crucial, we can change it later.